src/rl78.c

```diff
@@ -40,48 +48,105 @@ static void rl78_set_reset(port_handle_t fd, int mode, int value)
+        }
+        else
+        {
         printf("Using communication mode %u%s\n",
                (mode & (MODE_UART | MODE_RESET)) + 1,
                (mode & MODE_INVERT_RESET) ? " with RESET inversion" : "");
+        }   
     }
-    rl78_set_reset(fd, mode, 0);                            /* RESET -> 0 */
-    serial_set_txd(fd, 0);                                  /* TOOL0 -> 0 */
-    if (wait)
+    // Begin reset procedure
+    if (mode == 4) // sequencing for mode 5
     {
-        printf("Turn MCU's power on and press any key...");
-        wait_kbhit();
-        printf("\n");
+        //printf("Turn MCU's power on with RESET low. Press any key to continue...\n");
+        //wait_kbhit();
+        serial_set_txd(fd, 0);
+        printf("Press and hold RESET.\n");
+        int reset = serial_get_cts(fd);
+        if (reset == 1)
+        {
+            while (1) // while CTS/RESET is high, wait
+            {
+                reset = serial_get_cts(fd);
+                if (reset == 0)
+                {
+                    printf("Got reset press. Hang on...\n");
+                    usleep(100000);
+                    printf("Let go of RESET.\n");
+                    break;
+                }
+            }
+        }
+        
+        while (1) // while CTS/RESET is low, wait
+        {
+            reset = serial_get_cts(fd);
+            if (reset == 1)
+            {
+               printf("Got reset release. Attempting to set communication mode...\n");
+               break;
+            }
+        }
+        // Reset is now high
+        serial_flush(fd);
+        usleep(3000);
+        serial_set_txd(fd,1); // TOOL0 -> 1
+        usleep(1000);
+        serial_flush(fd);
+        
+    }
+    else if ((mode == 0) | (mode == 1) | (mode == 2))
+    { // sequencing for mode 1,2,3,4
+        rl78_set_reset(fd, mode, 0);                            /* RESET -> 0 */
+        serial_set_txd(fd, 0);                                  /* TOOL0 -> 0 */
+        if (wait)
+        {
+            printf("Turn MCU's power on and press any key...");
+            wait_kbhit();
+            printf("\n");
+        }
+        serial_flush(fd);
+        usleep(1000);
+        rl78_set_reset(fd, mode, 1);                            /* RESET -> 1 */
+        usleep(3000);
+        serial_set_txd(fd, 1);                                  /* TOOL0 -> 1 */
+        usleep(1000);
+        serial_flush(fd);
     }
-    serial_flush(fd);
-    usleep(1000);
-    rl78_set_reset(fd, mode, 1);                            /* RESET -> 1 */
-    usleep(3000);
-    serial_set_txd(fd, 1);                                  /* TOOL0 -> 1 */
-    usleep(1000);
-    serial_flush(fd);
     if (3 <= verbose_level)
     {
         printf("Send 1-byte data for setting mode\n");
     }
     serial_write(fd, &r, 1);
-    if (1 == communication_mode)
+    if ((1 == communication_mode) | (5 == communication_mode))
```

Comment:
I think the logic here is really whether it's a 1-Wire or 2-Wire serial.  Above around lines 63 and 69, we see this: `r = SET_MODE_1WIRE_UART;` or `r = SET_MODE_2WIRE_UART;`

I think it makes sense to determine serial type in one place, and use that determination in subsequent decisions.

read: Change this to `if (r == SET_MODE_1WIRE_UART)`

What's happening here is:  A 1 wire serial port is half duplex, so when you write a character, you also end up reading that same character.  This code is just to read that byte out of the buffer so it doesn't confuse the rest of the parser.